compiler/src/main/proto/spine/protodata/ast.proto

```diff
@@ -0,0 +1,256 @@
+
+    TypeName name = 1;
+
+    // The prefix of the type's URL.
+    //
+    // A type URL has the following structure: "<type URL prefix>/<type FQN>". For more info, see
+    // the definition of `google.protobuf.Any.type_url`.
+    //
+    string type_url_prefix = 2;
+
+    // The fields of this type.
+    //
+    // Does not include the fields belonging to a field group.
+    //
+    repeated Field field = 3;
+
+    // The field groups defined in this type.
+    repeated Group group = 4;
+
+    // Message-level options.
+    repeated Option option = 5;
+}
+
+// A primitive Protobuf type.
+enum PrimitiveType {
+
+    PT_UNKNOWN = 0;
+
+    TYPE_DOUBLE = 1;
+    TYPE_FLOAT = 2;
+    TYPE_INT64 = 3;
+    TYPE_UINT64 = 4;
+    TYPE_INT32 = 5;
+    TYPE_FIXED64 = 6;
+    TYPE_FIXED32 = 7;
+    TYPE_BOOL = 8;
+    TYPE_STRING = 9;
+    TYPE_BYTES = 10;
+    TYPE_UINT32 = 11;
+    TYPE_SFIXED32 = 12;
+    TYPE_SFIXED64 = 13;
+    TYPE_SINT32 = 14;
+    TYPE_SINT64 = 15;
+}
+
+// A Protobuf type.
+message Type {
+
+    oneof kind {
+
+        MessageType message = 1;
+
+        PrimitiveType primitive = 2;
+    }
+}
+
+// A name of a field in a message.
+message FieldName {
+
+    // Short name of the field.
+    string value = 1;
+}
+
+// A Protobuf message field.
+message Field {
+
+    FieldName name = 1;
+
+    // The type of the field.
+    Type type = 2;
+
+    // The type in which the field is declared.
+    MessageType declaring_type = 3;
+
+    // The field number.
+    //
+    // Not to be confused with the field's order to declaration.
+    //
+    // For example, in a message:
+    // ```
+    // message Foo {
+    //     string bar = 2;
+    //     int32 baz = 1;
+    // }
+    // ```
+    // Field `bar` has the number 2 and the field `baz` has the number 1.
+    //
+    uint32 number = 4;
+
+    // The field's order to declaration.
+    //
+    // Not to be confused with the field number.
+    //
+    // For example, in a message:
+    // ```
+    // message Foo {
+    //     string bar = 2;
+    //     int32 baz = 1;
+    // }
+    // ```
+    // Field `bar` has the order 0 and the field `baz` has the order 1.
+    //
+    uint32 order_of_declaration =54;
+
+    // The flag telling if the field is a list.
+    //
+    // A list is any `repeated` field which is not a `map`.
+    //
+    bool is_list = 6;
+
+    // The flag telling if the field is a map.
+    bool is_map = 7;
+
+    // If the field is a map, this is the type of the key.
+    //
+    // Otherwise, must be empty.
+    //
+    PrimitiveType map_key_type = 8;
+
+    // If this field is declared in a field group, the name of the group.
+    //
+    // Otherwise, must be empty.
+    //
+    GroupName group = 9;
+
+    // Field-level options.
+    repeated Option option = 10;
+}
+
+// Name of a group of fields.
+message GroupName {
+
+    // Name of the `oneof` block.
+    string value = 1;
+}
+
+// Group is a set of fields declared within one `oneof` block.
+message Group {
```

Comment:
Please consider renaming this to `OneofGroup` (and `GroupName` to `OneofName`). That would eliminate any potential confusions in the generated code, which would eventually happen with such a widespread word as "group".